HumanEvalLean/HumanEval106.lean

```diff
@@ -1,5 +1,207 @@
+  let mut ret : List Nat := []
+  for i in 1...=n do
+    if i % 2 = 0 then
+      let mut x := 1
+      for j in 1...=i do x := x * j
+      ret := x :: ret
+    else
+      let mut x := 0
+      for j in 1...=i do x := x + j
+      ret := x :: ret
+  return ret.reverse
+
+/-!
+## Tests
+-/
+
+example : f 5 = [1, 2, 6, 24, 15] := by native_decide
+example : f 7 = [1, 2, 6, 24, 15, 720, 28] := by native_decide
+example : f 1 = [1] := by native_decide
+example : f 3 = [1, 2, 6] := by native_decide
+
+/-!
+## Standard library wishlist
+
+The following lemmas halp with the verification and would be nice to have in the standard library.
+-/
+
+@[simp]
+theorem Std.PRange.length_toList_eq_size [UpwardEnumerable α] [LawfulUpwardEnumerable α]
+    [SupportsLowerBound sl α] [SupportsUpperBound su α] [BoundedUpwardEnumerable sl α]
+    [HasFiniteRanges su α] [RangeSize su α] [LawfulRangeSize su α]
+    {r : PRange ⟨sl, su⟩ α} :
+    r.toList.length = r.size := by
+  simp [PRange.toList, PRange.size]
+
+@[simp]
+theorem Nat.size_Rco {a b : Nat} :
+    (a...b).size = b - a := by
+  simp [Std.PRange.size, Std.Iterators.Iter.size, Std.Iterators.IteratorSize.size,
+    Std.PRange.Internal.iter, Std.Iterators.Iter.toIterM, Std.PRange.RangeSize.size]
+
+@[simp]
+theorem Nat.size_Rcc {a b : Nat} :
+    (a...=b).size = b + 1- a := by
+  simp [Std.PRange.size, Std.Iterators.Iter.size, Std.Iterators.IteratorSize.size,
+    Std.PRange.Internal.iter, Std.Iterators.Iter.toIterM, Std.PRange.RangeSize.size]
+
+theorem Std.PRange.getElem?_Rcx_eq [LE α] [UpwardEnumerable α] [LawfulUpwardEnumerable α]
+    [SupportsUpperBound su α] [LawfulUpwardEnumerableUpperBound su α]
+    [LawfulUpwardEnumerableLE α] [HasFiniteRanges su α]
+    {r : PRange ⟨.closed, su⟩ α} {i} :
+    r.toList[i]? = (UpwardEnumerable.succMany? i r.lower).filter (SupportsUpperBound.IsSatisfied r.upper) := by
+  induction i generalizing r
+  · rw [PRange.toList_eq_match, UpwardEnumerable.succMany?_zero]
+    simp only [Option.filter_some, decide_eq_true_eq]
+    split <;> simp
+  · rename_i n ih
+    rw [PRange.toList_eq_match]
+    simp only
+    split
+    · simp [LawfulUpwardEnumerable.succMany?_succ_eq_succ?_bind_succMany?]
+      cases hs : UpwardEnumerable.succ? r.lower
+      · rw [PRange.toList_eq_match]
+        simp [BoundedUpwardEnumerable.init?, hs]
+      · rw [toList_open_eq_toList_closed_of_isSome_succ? (by grind)]
+        rw [ih]
+        simp [hs]
+    · simp only [List.length_nil, Nat.not_lt_zero, not_false_eq_true, getElem?_neg]
+      cases hs : UpwardEnumerable.succMany? (n + 1) r.lower
+      · grind
+      · rename_i hl a
+        simp only [Option.filter_some, decide_eq_true_eq, right_eq_ite_iff]
+        have : UpwardEnumerable.LE r.lower a := ⟨n + 1, hs⟩
+        intro ha
+        exact hl.elim <| LawfulUpwardEnumerableUpperBound.isSatisfied_of_le r.upper _ _ ha this (α := α)
+
+theorem Std.PRange.isSome_succMany?_of_lt_length_toList [LE α] [UpwardEnumerable α]
```

Comment:
I know this isn't the place for design discussions around the range API, but this lemma does point out something interesting: there is a need to be able to ("noncomputably") talk about the size of a range, even if there is no `RangeSize` instance available because the size cannot be determined efficiently.

Reply:
Raising questions about the practicability of the API is a big benefit of human-eval-lean, so thank you for this question! I'm not sure how to solve this problem best right now, but I'll let that thought simmer in my head.